ANCS Support

cyberneel opened the pull request:
This PR Adds support for the Apple Notification Center Service (ANCS). It shows the Title and Subtitles as "Title - Subtitle" and the message is shown in the body of the InfiniTime notification. Accepting and Declining calls also worked.

The only "issue" right now is that you have to ReConnect after the first pair for ANCS discovery to go well. Maybe there is a way in the NimBLE stack to trigger ANCS Discovery after pairing change? Also, sometimes ANCS goes dormant and then works again. I think this could either be iOS throttling or 

> "ANCS is not guaranteed to always be present. As a result, the NC should look for and subscribe to the Service Changed characteristic of the GATT service in order to monitor for the potential publishing and unpublishing of the ANCS at any time."

cyberneel committed:
Started working on ANCS

cyberneel committed:
Started working on ANCS

cyberneel committed:
Started working on he client, which is the correct half that needs to be implemented.

cyberneel committed:
Started working on he client, which is the correct half that needs to be implemented.

cyberneel committed:
Formatting

cyberneel committed:
Formatting

cyberneel committed:
Added RTT logging

cyberneel committed:
Added RTT logging

cyberneel committed:
Added debugging through watch notifications

cyberneel committed:
Added debugging through watch notifications

cyberneel committed:
Removed encryption of battery and added required pairing on connect GAP event

cyberneel committed:
Removed encryption of battery and added required pairing on connect GAP event

cyberneel committed:
ANCS Partially works

Currently, the notification UID, Event ID and Category are shown.

cyberneel committed:
ANCS Partially works

Currently, the notification UID, Event ID and Category are shown.

cyberneel committed:
Started working on using Control Point and Data Source:

Data source is not working properly, the Characteristic and/or descriptor isn't being found correctly

cyberneel committed:
Started working on using Control Point and Data Source:

Data source is not working properly, the Characteristic and/or descriptor isn't being found correctly

cyberneel committed:
More debugging

cyberneel committed:
More debugging

cyberneel committed:
Got Datsource to somewhat work and notifications now show title.

Datasource seems to go off and on.

cyberneel committed:
Got Datsource to somewhat work and notifications now show title.

Datasource seems to go off and on.

cyberneel committed:
Fixed check to only let new notifications pass (maybe)

cyberneel committed:
Fixed check to only let new notifications pass (maybe)

cyberneel committed:
Proper Notification works.

Need to iron out some things like characteristics disappearing and reappearing.

cyberneel committed:
Proper Notification works.

Need to iron out some things like characteristics disappearing and reappearing.

cyberneel committed:
Fixed the event flag detection. Only shows a notification when it's new.

Also added constants to set lengths for the title, subtitle, and message.

cyberneel committed:
Fixed the event flag detection. Only shows a notification when it's new.

Also added constants to set lengths for the title, subtitle, and message.

cyberneel committed:
Fixed first letter in Subtitle being cut off

cyberneel committed:
Fixed first letter in Subtitle being cut off

liamcharger committed:
Improve message display

liamcharger committed:
Improve message display

cyberneel committed:
Updated notification, Title and Subtitle are in the orange test and message is in the body.

cyberneel committed:
Updated notification, Title and Subtitle are in the orange test and message is in the body.

cyberneel committed:
Implemented Incoming Calls to ANCS

Accepting and Declining works

cyberneel committed:
Implemented Incoming Calls to ANCS

Accepting and Declining works

cyberneel committed:
Turned off non important debugs

cyberneel committed:
Turned off non important debugs

cyberneel committed:
Formatted Code

cyberneel committed:
Formatted Code

liamcharger commented:
For reference, closes #910

Instructions: since an encrypted connection is required before ANCS is available, you’ll need to disconnect your watch from your companion app after updating, then reconnect. You’ll get a passkey on your watch to enter into the dialog on your iPhone. After bonding, toggle off and on your iPhone’s Bluetooth, then tap InfiniTime in the list of bonded devices. Finally, you’ll get an alert asking you to allow your watch to read notifications. Click allow, and now you should start receiving notifications!

Something to note: as of now, since using ANCS creates a connection through iOS directly instead of through the companion app, the app will not be able to reconnect to InfiniTime without removing the bond, which results in ANCS being disabled until the steps above are repeated. We’re working on fixing this ;)

liamcharger committed:
Remove RTT settings

liamcharger committed:
Remove RTT settings

cyberneel committed:
Removed more debug notifications

cyberneel committed:
Removed more debug notifications

cyberneel committed:
Added check for special characters and show unknown symbol.

cyberneel committed:
Added check for special characters and show unknown symbol.

cyberneel committed:
removed all debug notifications

cyberneel committed:
removed all debug notifications

cyberneel committed:
Ignore silent notifications

cyberneel committed:
Ignore silent notifications

cyberneel committed:
Make trunacated messages end in "..." to indicate that message has more on iOS

cyberneel committed:
Make trunacated messages end in "..." to indicate that message has more on iOS

minacode commented on src/components/ble/AppleNotificationCenterClient.h:
```diff
@@ -0,0 +1,134 @@
+        NegativeAction = (1 << 4)
+      };
+
+      struct AncsNotitfication {
```
typo

liamcharger replied:
Fixed

liamcharger committed:
Fix typos and remove unnecessary commented code

liamcharger committed:
Fix typos and remove unnecessary commented code

liamcharger committed:
Merge branch 'main' into ancs

mark9064 commented:
Does this implement subscribing to service changes? I think this is probably required given the ANCS can come and go at any time

I think the implementation of this is relatively simple, subscribe to the generic attribute service changed characteristic, and any time an indication is received service discovery should be performed again

cyberneel commented:
@mark9064 I didn't explicitly implement the service changed part, but in all my testing the watch was always able to recover from a disconnect and showed new notifications. So maybe there is something in the stack that handles this?

tituscmd committed:
Merge pull request #26 from InfiniTimeOrg/main

Alarm: Simplify alarm alerting screen (#2211)

mark9064 commented:
I don't believe it's handled elsewhere, but it could be? The text you quote at the top, about the service being registered/unregistered, isn't talking about bluetooth disconnects but actually the service itself disappearing from the list of services the apple device exposes, and then re-appearing at some later time. Maybe in practice apple devices never really do this (even though the spec allows them to), so you don't observe any issues. But I think it would be wise to have this base covered if the standard allows it

minacode commented:
How long did you test it? Could it be that the service disappears after a few hours?

liamcharger commented:
I'm able to use this for multiple days at a time without missing a notification. 

According to the logs, the watch is (very frequently) not able to subscribe to the data source, but always recovers before the next notification is sent.

liamcharger committed:
Merge branch 'main' into ancs

minacode commented:
Thank you for implementing this, @cyberneel!
I just typed the secure connection code into my iPhone for the first time, am already a little hyped from that and will test this PR over the next week :)

mark9064 commented on src/components/ble/AppleNotificationCenterClient.cpp:
```diff
@@ -0,0 +1,514 @@
+  auto isInFontDefinition = [](uint32_t codepoint) -> bool {
+    // Check if the codepoint falls into the specified font ranges or is explicitly listed
+    return (codepoint >= 0x20 && codepoint <= 0x7E) ||   // Printable ASCII
+           (codepoint >= 0x410 && codepoint <= 0x44F) || // Cyrillic
```
How come there's a Cyrillic check? I thought this wasn't supported by any of the fonts. More generally do we need some generic way to check this is within the font ranges? How does the existing notification client handle this?

mark9064 commented on src/components/ble/AppleNotificationCenterClient.cpp:
```diff
@@ -0,0 +1,514 @@
+  } else {
+    NRF_LOG_INFO("ANCS New alert subscribe ERROR");
+  }
+  if (isDescriptorFound == isControlDescriptorFound && isDescriptorFound == isDataDescriptorFound)
```
This is hard to read. What does it mean semantically to compare the values of `isDescriptorFound` and `isControlDescriptorFound` (for example)? Is the meaning the same if they're both `false` compared to when they're both `true`?

mark9064 reviewed (commented):
This is an enormous achievement, well done. There's a lot of code, so I have a lot of comments - no rush to getting to them all.

There are a few overall architectural bits that I've got questions about:
- There's a lot state being introduced. Is all of it being used? If possible, I think we should aim to store the minimum state required to show notification titles and text and allow calls to be accepted/declined.
- There's quite a lot of std::string usage, and I'm worried it might create a lot of heap pressure which may conflict with memory intensive watchfaces like G7710. Would it be possible to minimise memory usage by moving bytes from the BLE stack into the allocated notification object more directly? If not that's OK, but we should carefully assess the memory pressure possible.
- There seems to be an entire private storage of ANCS notifications inside the ANCS client. It'd be really great to just have one store of notifications, with that being the central notification controller. On a similar note, the extra ANCS ID being attached to each notification doesn't feel great. Maybe a UUID for each notification would be good though - perhaps refactoring towards a generic ID also usable by the other notification clients could be good? What do you think?

This review isn't complete as there's probably more things to discuss, but hopefully it's a good starting point.

Great job with this implementing this :)

cyberneel commented on src/components/ble/AppleNotificationCenterClient.cpp:
```diff
@@ -0,0 +1,514 @@
+    bool silent = (eventFlags & static_cast<uint8_t>(EventFlags::Silent)) != 0;
+    // bool important = eventFlags & static_cast<uint8_t>(EventFlags::Important);
+    bool preExisting = (eventFlags & static_cast<uint8_t>(EventFlags::PreExisting)) != 0;
+    // bool positiveAction = eventFlags & static_cast<uint8_t>(EventFlags::PositiveAction);
```
ANCS provides tags for what type of notification it is sending, and I wasn't using those at the time.

cyberneel commented on src/displayapp/screens/Notifications.h:
```diff
@@ -38,20 +39,25 @@ namespace Pinetime {
             return running;
           }
 
+          uint32_t ancsUid = 0;
```
The ANCS system has a separate endpoint for actions. So I can only accept calls for example if I reference to the uuid of the call notification and send a positive command. So, this id is different than a generic id for internal use, it is required for ANCS to know which notifications I am talking about.

cyberneel replied:
Apparently, on Android it is a generic signal that sends to the device. But iOS requires the signal to be associated with a specific notification to affect it.

minacode commented:
Yes, it is really impressive that you did it. Thank you!

Yesterday I found that we have no-device support for German umlauts öäü and other such characters. Since the notifications come directly from iOS, we have to discuss how far we can and want to go on the watch. 

But that would be no blocking issue for this PR from my side as long as the BLE service works fine otherwise. So maybe discuss this in the linked issue instead of here.

minacode commented:
My test over the last week was very positive. Great job!
Feature-wise I would merge it as is, but haven't looked at the code in detail yet and will hopefully do so in the next days.

minacode commented on src/components/ble/AppleNotificationCenterClient.cpp:
```diff
@@ -0,0 +1,514 @@
+    bool incomingCall = ancsNotif.uuid != 0 && ancsNotif.category == static_cast<uint8_t>(Categories::IncomingCall);
+
+    if (!incomingCall) {
+      if (titleSize >= maxTitleSize) {
+        decodedTitle.resize(maxTitleSize - 3);
+        decodedTitle += "...";
+        if (!decodedSubTitle.empty()) {
+          decodedTitle += " - ";
+        } else {
+          decodedTitle += "-";
+        }
```
if `decodedTitle` is longer than `maxTitleSize`, then you resize it to 3 chars less than max. Then you add `"..."`, which makes it full again, and then you add `" - "` or even `-`. To me, it seems like `decodedTitle` is again too long after that.

minacode commented on src/components/ble/AppleNotificationCenterClient.cpp:
```diff
@@ -0,0 +1,514 @@
+void AppleNotificationCenterClient::OnNotification(ble_gap_event* event) {
+  if (event->notify_rx.attr_handle == notificationSourceHandle || event->notify_rx.attr_handle == notificationSourceDescriptorHandle) {
+    NRF_LOG_INFO("ANCS Notification received");
+    uint8_t eventId;
+    uint8_t eventFlags;
+    uint8_t category;
+    uint8_t categoryCount;
+    uint32_t notificationUuid;
+
+    os_mbuf_copydata(event->notify_rx.om, 0, 1, &eventId);
+    os_mbuf_copydata(event->notify_rx.om, 1, 1, &eventFlags);
+    os_mbuf_copydata(event->notify_rx.om, 2, 1, &category);
+    os_mbuf_copydata(event->notify_rx.om, 3, 1, &categoryCount);
+    os_mbuf_copydata(event->notify_rx.om, 4, 4, &notificationUuid);
```
Maybe gcc is smart enough to get this, but I will aks anyway :smile:
You create those local variables only to insert them to a notification. Couldn't you cut the local vairables and `os_mbuf_copydata` directly into the `ancsNotif`?

minacode commented on src/components/ble/AppleNotificationCenterClient.cpp:
```diff
@@ -0,0 +1,514 @@
+    ancsNotif.category = category;
+    ancsNotif.uuid = notificationUuid;
+
+    while (notifications.size() > 6) {
```
Do we have a less magical way to express this 6?

minacode commented on src/components/ble/AppleNotificationCenterClient.cpp:
```diff
@@ -0,0 +1,514 @@
+    }
+
+    // Request ANCS more info
+    uint8_t titleSize = maxTitleSize + 4;
+    uint8_t subTitleSize = maxSubtitleSize + 4;
+    uint8_t messageSize = maxMessageSize + 4;
```
Can you add a comment here, where that `+ 4` comes from?

minacode commented on src/components/ble/AppleNotificationCenterClient.cpp:
```diff
@@ -0,0 +1,514 @@
+
+    NotificationManager::Notification notif;
+    notif.ancsUid = notificationUid;
+    std::string notifStr;
+
+    if (incomingCall) {
+      notifStr += "Incoming Call:";
+      notifStr += decodedTitle;
+      notifStr += "\n";
+      notifStr += decodedSubTitle;
+    } else {
+      notifStr += decodedTitle;
```
The first `notifStr += ...` in the if and else can be an `notifStr = ...` without `+`, because `notifStr` is new.

minacode commented on src/components/ble/AppleNotificationCenterClient.cpp:
```diff
@@ -0,0 +1,514 @@
+    }
+
+    // Adjust notification if too long
+    if (notifStr.size() > 100) {
+      notifStr.resize(97);
```
`NotificationManager` has an attribute `MessageSize` which encodes this 100.
Please use this and replace the two magic numbers.

minacode commented on src/components/ble/AppleNotificationCenterClient.cpp:
```diff
@@ -0,0 +1,514 @@
+      notifStr += "...";
+    }
+
+    notif.message = std::array<char, 101> {};
```
Replace magic number: `101` -> `MessageSize + 1`

minacode commented on src/components/ble/AppleNotificationCenterClient.cpp:
```diff
@@ -0,0 +1,514 @@
+    notif.message = std::array<char, 101> {};
+    std::strncpy(notif.message.data(), notifStr.c_str(), notif.message.size() - 1);
+
+    if (incomingCall)
+      notif.message[13] = '\0'; // Separate Title and Message
+    else if (!decodedSubTitle.empty())
+      notif.message[titleSize + subTitleSize] = '\0'; // Separate Title and Message
+    else
+      notif.message[titleSize - 1] = '\0'; // Separate Title and Message
```
Please use braces here, just to be safe.

minacode reviewed (changes requested):
Again, thank you for your contribution! I have reviewed the first half of your code and found some smaller issues. Overall, the code is written well and one can see that you figured out some clever things to get this feature up and running.
My feedback primarily focuses on two areas:
1. Seeing if we can cut some needed memory.
2. Reducing the amount of magic numbers by using constants or adding some explanations.

tituscmd commented:
> I don't know if this helps, but I've been running this branch on my PineTime and I have been encountering an occasional crash and reboot. I don't know if it's related to the changes here. Is there anything I can provide that would help the development efforts here? Some sort of log in a version of InfiniLink maybe?

Can confirm this on a fork of mine including ANCS

joelghill commented:
I'm more than willing to provide logs and debug information to help resolve the issue. Does the watch record any logs, or is that something a companion app can provide?

cyberneel commented:
Hello everyone,

I've been very busy since I made this PR with school and other commitments.

Most of the base for my code was adapted from the existing notification service meant for android. So there are probably many places I could have optimized.

I plan on responding to code reviews in the coming weeks to give everyone a better understanding of the code. This could help narrow down possible cause of the reboots.

Thank you everyone for your support!

minacode commented:
> Thank you everyone for your support!

Thank you for your support! ;)

minacode commented:
I hope I can find some time next weekend for further review. It's been a while. Sorry for that.

tituscmd committed:
Merge branch 'InfiniTimeOrg:main' into main

tituscmd committed:
Merge branch 'InfiniTimeOrg:main' into main

minacode commented:
I am still running this branch all the time. I noticed one crash in the meantime, although more might have happened, because I wouldn't have noticed as long as the watch reconnected and reset its time correctly after rebooting.

I am overall very positive, still have to review the last half of the code (sorry), but would say that this PR is totally usable and I would already in this state recommend it to anyone using a PineTime with iOS.

Let's both find the time to review it (reviews from others who want to help are welcome as well 🙂) and this is on a good way to be merged.

Also please update this branch to the current main.

liamcharger commented:
> Also please update this branch to the current main.

I can take care of that :)

liamcharger committed:
Merge branch 'main' into ancs

cyberneel commented:
@minacode Thank you! I am on break from school now so I have some time. I can start looking at the code review and make comments.

I'll update the branch soon.

cyberneel commented on src/displayapp/fonts/fonts.json:
```diff
@@ -3,7 +3,7 @@
          {
             "file": "JetBrainsMono-Bold.ttf",
-            "range": "0x20-0x7e, 0x410-0x44f, 0xB0"
+            "range": "0x20-0x7e, 0x410-0x44f, 0xB0, 0xBF, 0xFFFD"
```
I think it looks pretty nice. Here is a photo:

![IMG_3391](https://github.com/user-attachments/assets/75b6d12c-e4bc-46d0-8bc4-8fda74e13c82)

cyberneel commented on src/displayapp/fonts/fonts.json:
```diff
@@ -3,7 +3,7 @@
          {
             "file": "JetBrainsMono-Bold.ttf",
-            "range": "0x20-0x7e, 0x410-0x44f, 0xB0"
+            "range": "0x20-0x7e, 0x410-0x44f, 0xB0, 0xBF, 0xFFFD"
```
I think it was the "¿" character, but since it could be used as punctuation, I then replaced it with "�".

Can be removed I think.

cyberneel reviewed (commented):
I responded to @mark9064 comments on the code, I'll get to the other comments soon.

cyberneel reviewed (commented):
I think I addressed all the comments, just have to make the code updates now.

cyberneel committed:
Made codebase changes based on PR Comments

https://github.com/InfiniTimeOrg/InfiniTime/pull/2217/files/de63dce4ce5ef9b796013a6a9150d90b81fc3981#diff-a57dcf730dd40f33f9081712a57f33d39d6a73096f630df41deb58f5032e1870

cyberneel committed:
Made codebase changes based on PR Comments

https://github.com/InfiniTimeOrg/InfiniTime/pull/2217/files/de63dce4ce5ef9b796013a6a9150d90b81fc3981#diff-a57dcf730dd40f33f9081712a57f33d39d6a73096f630df41deb58f5032e1870

cyberneel committed:
Merge branch 'main' into ancs

tituscmd committed:
Merge branch 'InfiniTimeOrg:main' into main

tituscmd committed:
Merge branch 'InfiniTimeOrg:main' into main

cyberneel committed:
Merge branch 'InfiniTimeOrg:main' into ancs

cyberneel committed:
Added security initiate back for ancs to work.

cyberneel committed:
Added security initiate back for ancs to work.

cyberneel committed:
Updated checks to prevent duplicate notification triggers, NEEDS testing but not issues so far.

cyberneel committed:
Updated checks to prevent duplicate notification triggers, NEEDS testing but not issues so far.

cyberneel committed:
Check for silent before ringing notification

cyberneel committed:
Check for silent before ringing notification

cyberneel committed:
Fix typo in ping check

cyberneel committed:
Fix typo in ping check

liamcharger commented:
I've noticed this as well...

Nothing changes when they're sent again, like any flags, etc.?

cyberneel commented:
I have to test more, but this ping happens in the datasource handler. And the datasource doesn't provide flag information.

The notification source handler has a return statement if it's not a new notification source it doesn't request datasource at all.

I think possibly datasource is sending multiple times some times for some reason. I'll see if check if uuid exists in notification manager fixes it.

cyberneel committed:
ignore notifications not in list

cyberneel committed:
ignore notifications not in list

cyberneel committed:
Added a session uuid cache to try to help with duplicate pings

cyberneel committed:
Added a session uuid cache to try to help with duplicate pings

cyberneel committed:
clear session uids on reset

cyberneel committed:
clear session uids on reset

cyberneel committed:
removed extra array

cyberneel committed:
removed extra array

tituscmd committed:
Merge branch 'InfiniTimeOrg:main' into main

tituscmd commented:
> I don't know if this helps, but I've been running this branch on my PineTime and I have been encountering an occasional crash and reboot. I don't know if it's related to the changes here. Is there anything I can provide that would help the development efforts here? Some sort of log in a version of InfiniLink maybe?

For anyone experiencing this as well, I've done some testing and found a way to, in my case, 100% reliably cause a watchdog reset.

**Step 1**: Do all the ANCS connection stuff so that your watch mirrors your phone's notifications like usual
**Step 2**: Send lots of very long messages. I tried it with roundabout 15 messages, 2000 characters each message. I'm not sure if this is necessarily needed, since the watch only stores 5 notifications and each one gets cut off after some hard-coded character limit, but that's what I tested it with.
**Step 3**: Go into your iPhone's Bluetooth settings and toggle it off and on pretty frequently. In my case, once the watch showed the little Bluetooth symbol in the status icons, I turned Bluetooth off and when it was gone, I turned it back on. I did that just to make sure it was definitely reconnecting each time.
**Step 4**: After about 15 - 20 seconds, the watch will freeze and cause a watchdog reboot

Now, I'm not an expert in ble logic, infact I don't really know anything about it, so this is where my side of the info stops. I hope someone else that knows more about it can profit from this test and figure out a fix.

tituscmd commented:
After further investigation, notifications in the Notification Center don't matter. It's just the reconnection that's causing it

minacode commented:
There are still crashes. The good news is that InfiniTime is robust enough that you won't notice them if you don't look at the screen during the reboot that happens after a crash.
While we are not ready for merging yet, I am already running this branch for a long time and it does the job well enough to use it in daily life. So try it out! 🙂

tituscmd commented:
I actually found a solution to the crashes. It's not perfect, but it works! Let me get a PR going real quick...

tituscmd committed:
Merge branch 'ancs' of https://github.com/cyberneel/InfiniTime into ancs-fix-crashes

tituscmd committed:
add new helper method

tituscmd committed:
add another condition for connecting

tituscmd committed:
add in header as well

tituscmd commented:
> There are still crashes. The good news is that InfiniTime is robust enough that you won't notice them if you don't look at the screen during the reboot that happens after a crash. While we are not ready for merging yet, I am already running this branch for a long time and it does the job well enough to use it in daily life. So try it out! 🙂

I actually found a solution to the crashes! It's not perfect, but it works :D
Let me get a PR going real quick...

liamcharger commented:
Sigh, meant to merge instead of rebase

Never again will I use the github.com buttons

liamcharger committed:
Fix formatting

liamcharger commented:
@Crefix notifications are still sent multiple times in some cases and the simulator build needs to be fixed

cyberneel commented:
Also, I had a fix for the simulator build locally somewhere. But I do remember it being a simple fix.

tituscmd commented:
> @liamcharger is this with the PR that @tituscmd proposed?

While on topic, sorry for not getting to that better solution we talked about in that PR yet. It's exam season so I tend to be quite busy with school. I think for now the PR is fine as it stands since it doesn't make the watch crash anymore. If that PR makes this one more likely to be merged, I think we can bother with an even better solution later. How does that sound to you?

tituscmd committed:
Merge branch 'ancs' into ancs-fix-crashes

cyberneel committed:
Merge pull request #19 from InfiniBros/ancs-fix-crashes

Fix race condition in ANCS discovery by adding MaybeFinishDiscovery helper

NOTE: Going to merge this for now, and I will then add the functionality I talk about [here](https://github.com/cyberneel/InfiniTime/pull/19#pullrequestreview-3328664031)

cyberneel commented:
> > I haven't come into any instances where notifications don't show up yet.
> 
> If by this you mean notifications appear on the phone but not on the watch, I have indeed had this happen

I see, I am trying to figure out behavior. Could you maybe note symptoms of when these 2 glitches happen next time?

tituscmd commented:
> I see, I am trying to figure out behavior. Could you maybe note symptoms of when these 2 glitches happen next time?

It's very difficult, if at all possible, to figure out a cause. I have yet to find any correlation at all...

minacode commented:
@cyberneel, can you gather a checkbox-list of all current issues in the main description?

liamcharger committed:
Merge branch 'main' into ancs

liamcharger commented:
@mark9064 can we get a workflow approval? 

Trying to test something out, but I can't get an environment set up for compiling right now ;)

mark9064 commented:
Triggered :)

jgonggrijp commented:
I know C++ and I am an experienced reviewer, but I have zero experience with firmware programming, nor with Apple-specific APIs. Would it help if I reviewed https://github.com/cyberneel/InfiniTime/pull/20?
